### PR TITLE
Issue 30213: Appending RolloutMetrics to sampler outputs should happen after(!) all callbacks (such that custom metrics for last obs are still included)

### LINKED ISSUE #30213: [RLLib] Customized metrics is missing the last observation
### What happened + What you expected to happen

Please correct me if I'm wrong.

In current logic, `outputs.append(RolloutMetrics)` is called before `on_episode_step` and `on_episode_end`, snippets shown below. So that the last obs' customized metrics never gets reported.

Screen Shot 2022-11-12 at 01 31 13
Screen Shot 2022-11-12 at 01 34 22

### Versions / Dependencies

2.0.0

### Reproduction script

As above

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by sven1977**
Signed-off-by: sven1977 <svenmika1977@gmail.com>

Issue 30213: Appending RolloutMetrics to sampler outputs should happen after(!) all callbacks (such that custom metrics for last obs are still included)

Closes #30213

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/sampler.py [hunk 1]
```diff
@@ -968,9 +944,6 @@ def _process_observations(
                     build=episode.is_faulty or not multiple_episodes_in_batch,
                 )
             if not episode.is_faulty:
-                if ma_sample_batch:
```
Comment (ArturNiederfahrenhorst) [inline rllib/evaluation/sampler.py:971]: Without going into depth here: This looks where only returning non-fauly episodes before and now return all episodes. Is that intended?

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Only have one concern!

File: rllib/evaluation/sampler.py [hunk 2]
```diff
@@ -995,6 +968,40 @@ def _process_observations(
...
+                            episode.media,
+                        )
+                    )
+            else:
+                # Add metrics about a faulty episode.
+                outputs.append(RolloutMetrics(episode_faulty=True))
+
+            # Only after the RolloutMetrics were appended, append the collected sample
+            # batch, if any.
+            if not episode.is_faulty and ma_sample_batch:
```
Comment (sven1977) [inline rllib/evaluation/sampler.py:1002]: Actually, I think it's fine. This is the place where the append got moved to and it only happens if the episode is NOT faulty and there is a MA sample batch, which are the exact two criteria from before this change.

File: rllib/evaluation/sampler.py (hunk 1 above)
Comment (sven1977) [reply rllib/evaluation/sampler.py:971]: See my comment below on the line of code where this logic was moved to.

File: rllib/evaluation/sampler.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/sampler.py:1002]: Huh, sorry.

**Review by ArturNiederfahrenhorst — APPROVED** [review]
lgtm, thanks!

**Comment by sven1977** [comment]
Thanks for your review @ArturNiederfahrenhorst !

**Pull request merged**